java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.search;
+
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * Search for the range of a particular element in the target vector.
+ */
+public class VectorRangeSearcher {
+
+  /**
+   * Result returned when a search fails.
+   */
+  public static final int SEARCH_FAIL_RESULT = -1;
+
+  /**
+   * Search for the first occurrence of an element.
+   * The search is based on the binary search algorithm. So the target vector must be sorted.
+   * @param targetVector the vector from which to perform the search.
+   * @param comparator the criterion for the comparison.
+   * @param keyVector the vector containing the element to search.
+   * @param keyIndex the index of the search key in the key vector.
+   * @param <V> the vector type.
+   * @return the index of the first matched element if any, and -1 otherwise.
+   */
+  public static <V extends ValueVector> int getFirstMatch(
+          V targetVector, VectorValueComparator<V> comparator, V keyVector, int keyIndex) {
+    comparator.attachVectors(keyVector, targetVector);
+
+    int low = 0;
+    int high = targetVector.getValueCount() - 1;
+
+    while (low <= high) {
+      int mid = low + (high - low) / 2;
+      int result = comparator.compare(keyIndex, mid);
+      if (result < 0) {
+        // the key is smaller
+        high = mid - 1;
+      } else if (result > 0) {
+        // the key is larger
+        low = mid + 1;
+      } else {
+        // the key equals the mid value, find the lower bound by going left-ward.
+

```

Comment:
The following requires less branching. I haven't validated if it properly works, so be cautious (or add more test!).

```
    int low = 0;
    int high = targetVector.getValueCount() - 1;
    int ret = SEARCH_FAIL_RESULT;

     while (low <= high) {
      int mid = low + (high - low) / 2;
      int result = comparator.compare(keyIndex, mid);
      if (result < 0) {
        // the key is smaller
        high = mid - 1;
      } else if (result > 0) {
        // the key is larger
        low = mid + 1;
      } else {
        // A match is found, stash it.
        ret = mid;
        // Binary search will continue by clamping on the left
        high = mid - 1; 
      }
     }
   return ret;
```

Reply:
@fsaintjacques Thanks a lot for your comments, and the improved algorithm.

I think to prove the correctness of the algorithm, we need to maintain some loop invariants. Suppose we call the boundary point (lower or upper bound) the bp, the following invariants must hold throughout:

1. low <= bp
2. high >= bp

It seems in some cases, this statement may break the second invariant (Maybe I am wrong)?

high = mid - 1

Anyway, it is interesting question. Would you please give more details about the reasoning behind your algorithm?